File: Data Structures & Algorithms/house-robber-iii/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    rob(root) {
        const dfs = (node) => {
            if (!node) {
                return [0, 0];
            }

            const left = dfs(node.left);
            const right = dfs(node.right);

            const withRoot = node.val + left[1] + right[1];
            const withoutRoot = Math.max(...left) + Math.max(...right);

            return [withRoot, withoutRoot];
        }

        return Math.max(...dfs(root));

    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        return dfs(root, Integer.MIN_VALUE);
    }

    private int dfs(TreeNode root, int maxSoFar){
        if(root == null) return 0;

        int count = root.val >= maxSoFar ? 1 : 0;

        int newMax = Math.max(maxSoFar, root.val);

        count += dfs(root.left, newMax);
        count += dfs(root.right, newMax);

        return count;
    }
}
